{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from process_helper import get_mageck_command"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# IntalaciÃ³n Mageck\n",
    "\n",
    "\"Reads were aligned to the appropriate reference library using MAGeCK version 0.5.9.2\"\n",
    "```bash\n",
    "conda install -c bioconda mageck=0.5.9.2\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For CD4 supplemental\n",
    "CRISPR_type = 'CRISPRa'\n",
    "cell_types = 'CD4'\n",
    "cytokine = 'TNFa'\n",
    "study_type = 'supplemental'\n",
    "input_folder = 'input_data'\n",
    "count_filename = f'{input_folder}/{CRISPR_type}.{cell_types}_count.{study_type}.txt'\n",
    "nt_filename = f'{input_folder}/{CRISPR_type}.{cell_types}_NO-TARGET.{study_type}.txt'\n",
    "out_filename = f'data_out/{CRISPR_type}.{cytokine}.{cell_types}.{study_type}'\n",
    "donor_1 = 15\n",
    "donor_2 = 16\n",
    "mageck_command = get_mageck_command(\n",
    "    count_filename,\n",
    "    donor_1,\n",
    "    donor_2,\n",
    "    cytokine,\n",
    "    out_filename,\n",
    "    nt_filename\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  @ Wed, 09 Oct 2024 18:02:12: Parameters: /home/julian/miniconda3/envs/mageckenv2/bin/mageck test -k input_data/CRISPRa.CD4_count.supplemental.txt -t Donor15_TNFa_high,Donor16_TNFa_high -c Donor15_TNFa_low,Donor16_TNFa_low -n data_out/CRISPRa.TNFa.CD4.supplemental --norm-method none --paired --control-sgrna input_data/CRISPRa.CD4_NO-TARGET.supplemental.txt --keep-tmp \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:12: Welcome to MAGeCK v0.5.9.2. Command: test \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:12: Loading count table from input_data/CRISPRa.CD4_count.supplemental.txt  \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:12: Processing 1 lines.. \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:13: Processing 100001 lines.. \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:13: Loaded 113238 records. \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:13: Loading R template file: /home/julian/miniconda3/envs/mageckenv2/lib/python3.7/site-packages/mageck/plot_template.RTemplate. \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:13: Loading R template file: /home/julian/miniconda3/envs/mageckenv2/lib/python3.7/site-packages/mageck/plot_template_indvgene.RTemplate. \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:13: Loading Rnw template file: /home/julian/miniconda3/envs/mageckenv2/lib/python3.7/site-packages/mageck/plot_template.Rnw. \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:13: Treatment samples:Donor15_TNFa_high,Donor16_TNFa_high \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:13: Treatment sample index:2,6 \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:13: Control samples:Donor15_TNFa_low,Donor16_TNFa_low \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:13: Control sample index:3,7 \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:13: Final size factor: 1.0 1.0 1.0 1.0 \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:14: The updated negative control ID file is:data_out/CRISPRa.TNFa.CD4.supplemental.negctrl_ids.txt \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:15: Detecting outliers from variance estimation.. Avg read count:[142.5], Var: [16980.423754277515] \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:15: Skipping 1289 sgRNAs from variance calculation because of their extreme high counts (> 4* STD (mean counts) ). \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:15: Before RRA, 100 sgRNAs are removed with zero counts in both group(s). \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:15: Use qnorm to reversely calculate sgRNA scores ... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:19: Running command: RRA -i data_out/CRISPRa.TNFa.CD4.supplemental.plow.txt -o data_out/CRISPRa.TNFa.CD4.supplemental.gene.low.txt -p 0.24203979220412059 --control data_out/CRISPRa.TNFa.CD4.supplemental.negctrl_ids.txt --skip-gene NA --skip-gene na  \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26: Command message: \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   Welcome to RRA v 0.5.9. \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   1985 control sequences loaded. \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   Skipping gene NA for permutation ... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   Skipping gene na for permutation ... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   Reading input file... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   Summary: 226376 sgRNAs, 18931 genes, 1 lists; skipped sgRNAs:0 \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   Computing lo-values for each group... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   Skipping gene NO-TARGET for permutation ... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   Warning: sgRNA  not found in the ranked list.  \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   Computing false discovery rate... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   Total # control sgRNAs: 1984 \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   Permuting genes with 2 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   Permuting genes with 4 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   Permuting genes with 6 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   Permuting genes with 8 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   Permuting genes with 9 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   Permuting genes with 10 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   Permuting genes with 11 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   Permuting genes with 12 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   Permuting genes with 24 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   Number of genes under FDR adjustment: 18931 \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   Saving to output file... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   Suppressing the output of gene NO-TARGET since it is negative ontrol genes. \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:   RRA completed. \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26:    \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26: End command message. \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:26: Running command: RRA -i data_out/CRISPRa.TNFa.CD4.supplemental.phigh.txt -o data_out/CRISPRa.TNFa.CD4.supplemental.gene.high.txt -p 0.38728045375834896 --control data_out/CRISPRa.TNFa.CD4.supplemental.negctrl_ids.txt --skip-gene NA --skip-gene na  \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35: Command message: \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   Welcome to RRA v 0.5.9. \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   1985 control sequences loaded. \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   Skipping gene NA for permutation ... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   Skipping gene na for permutation ... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   Reading input file... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   Summary: 226376 sgRNAs, 18931 genes, 1 lists; skipped sgRNAs:0 \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   Computing lo-values for each group... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   Skipping gene NO-TARGET for permutation ... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   Warning: sgRNA  not found in the ranked list.  \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   Computing false discovery rate... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   Total # control sgRNAs: 1984 \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   Permuting genes with 2 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   Permuting genes with 4 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   Permuting genes with 6 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   Permuting genes with 8 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   Permuting genes with 9 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   Permuting genes with 10 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   Permuting genes with 11 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   Permuting genes with 12 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   Permuting genes with 24 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   Number of genes under FDR adjustment: 18931 \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   Saving to output file... \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   Suppressing the output of gene NO-TARGET since it is negative ontrol genes. \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:   RRA completed. \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35:    \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35: End command message. \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:35: Loading top 10 genes from data_out/CRISPRa.TNFa.CD4.supplemental.gene.low.txt: LAT2,MAP4K1,LYN,MUC1,IKZF3,INPPL1,ARHGDIB,PITX1,SLA2,TOX \n",
      "INFO  @ Wed, 09 Oct 2024 18:02:37: Loading top 10 genes from data_out/CRISPRa.TNFa.CD4.supplemental.gene.high.txt: CD28,VAV1,DEF6,CNR2,OTUD7B,LCP2,TNFRSF9,DEPDC7,GPR84,P2RY14 \n"
     ]
    }
   ],
   "source": [
    "! {mageck_command}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For CD8 \n",
    "CRISPR_type = 'CRISPRa'\n",
    "cell_types = 'CD8'\n",
    "cytokine = 'IFNG'\n",
    "study_type = 'orig'\n",
    "input_folder = 'input_data'\n",
    "count_filename = f'{input_folder}/{CRISPR_type}.{cell_types}_count.{study_type}.txt'\n",
    "nt_filename = f'{input_folder}/{CRISPR_type}.{cell_types}_NO-TARGET.{study_type}.txt'\n",
    "out_filename = f'data_out/{CRISPR_type}.{cytokine}.{cell_types}.{study_type}'\n",
    "donor_1 = 1\n",
    "donor_2 = 2\n",
    "mageck_command = get_mageck_command(\n",
    "    count_filename,\n",
    "    donor_1,\n",
    "    donor_2,\n",
    "    cytokine,\n",
    "    out_filename,\n",
    "    nt_filename\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  @ Wed, 09 Oct 2024 18:03:55: Parameters: /home/julian/miniconda3/envs/mageckenv2/bin/mageck test -k input_data/CRISPRa.CD8_count.orig.txt -t Donor1_IFNG_high,Donor2_IFNG_high -c Donor1_IFNG_low,Donor2_IFNG_low -n data_out/CRISPRa.IFNG.CD8.orig --norm-method none --paired --control-sgrna input_data/CRISPRa.CD8_NO-TARGET.orig.txt --keep-tmp \n",
      "INFO  @ Wed, 09 Oct 2024 18:03:55: Welcome to MAGeCK v0.5.9.2. Command: test \n",
      "INFO  @ Wed, 09 Oct 2024 18:03:55: Loading count table from input_data/CRISPRa.CD8_count.orig.txt  \n",
      "INFO  @ Wed, 09 Oct 2024 18:03:55: Processing 1 lines.. \n",
      "INFO  @ Wed, 09 Oct 2024 18:03:55: Processing 100001 lines.. \n",
      "INFO  @ Wed, 09 Oct 2024 18:03:55: Loaded 113238 records. \n",
      "INFO  @ Wed, 09 Oct 2024 18:03:55: Loading R template file: /home/julian/miniconda3/envs/mageckenv2/lib/python3.7/site-packages/mageck/plot_template.RTemplate. \n",
      "INFO  @ Wed, 09 Oct 2024 18:03:55: Loading R template file: /home/julian/miniconda3/envs/mageckenv2/lib/python3.7/site-packages/mageck/plot_template_indvgene.RTemplate. \n",
      "INFO  @ Wed, 09 Oct 2024 18:03:55: Loading Rnw template file: /home/julian/miniconda3/envs/mageckenv2/lib/python3.7/site-packages/mageck/plot_template.Rnw. \n",
      "INFO  @ Wed, 09 Oct 2024 18:03:55: Treatment samples:Donor1_IFNG_high,Donor2_IFNG_high \n",
      "INFO  @ Wed, 09 Oct 2024 18:03:55: Treatment sample index:2,8 \n",
      "INFO  @ Wed, 09 Oct 2024 18:03:55: Control samples:Donor1_IFNG_low,Donor2_IFNG_low \n",
      "INFO  @ Wed, 09 Oct 2024 18:03:55: Control sample index:1,7 \n",
      "INFO  @ Wed, 09 Oct 2024 18:03:56: Final size factor: 1.0 1.0 1.0 1.0 \n",
      "INFO  @ Wed, 09 Oct 2024 18:03:56: The updated negative control ID file is:data_out/CRISPRa.IFNG.CD8.orig.negctrl_ids.txt \n",
      "INFO  @ Wed, 09 Oct 2024 18:03:57: Detecting outliers from variance estimation.. Avg read count:[69.0], Var: [1438.913826029363] \n",
      "INFO  @ Wed, 09 Oct 2024 18:03:57: Skipping 897 sgRNAs from variance calculation because of their extreme high counts (> 4* STD (mean counts) ). \n",
      "INFO  @ Wed, 09 Oct 2024 18:03:58: Before RRA, 113 sgRNAs are removed with zero counts in both group(s). \n",
      "INFO  @ Wed, 09 Oct 2024 18:03:58: Use qnorm to reversely calculate sgRNA scores ... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:01: Running command: RRA -i data_out/CRISPRa.IFNG.CD8.orig.plow.txt -o data_out/CRISPRa.IFNG.CD8.orig.gene.low.txt -p 0.31688924426695175 --control data_out/CRISPRa.IFNG.CD8.orig.negctrl_ids.txt --skip-gene NA --skip-gene na  \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10: Command message: \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   Welcome to RRA v 0.5.9. \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   1985 control sequences loaded. \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   Skipping gene NA for permutation ... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   Skipping gene na for permutation ... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   Reading input file... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   Summary: 226363 sgRNAs, 18931 genes, 1 lists; skipped sgRNAs:0 \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   Computing lo-values for each group... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   Skipping gene NO-TARGET for permutation ... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   Warning: sgRNA  not found in the ranked list.  \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   Computing false discovery rate... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   Total # control sgRNAs: 1984 \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   Permuting genes with 2 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   Permuting genes with 4 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   Permuting genes with 6 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   Permuting genes with 8 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   Permuting genes with 9 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   Permuting genes with 10 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   Permuting genes with 11 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   Permuting genes with 12 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   Permuting genes with 24 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   Number of genes under FDR adjustment: 18931 \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   Saving to output file... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   Suppressing the output of gene NO-TARGET since it is negative ontrol genes. \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:   RRA completed. \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10:    \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10: End command message. \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:10: Running command: RRA -i data_out/CRISPRa.IFNG.CD8.orig.phigh.txt -o data_out/CRISPRa.IFNG.CD8.orig.gene.high.txt -p 0.25166657095019945 --control data_out/CRISPRa.IFNG.CD8.orig.negctrl_ids.txt --skip-gene NA --skip-gene na  \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17: Command message: \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   Welcome to RRA v 0.5.9. \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   1985 control sequences loaded. \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   Skipping gene NA for permutation ... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   Skipping gene na for permutation ... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   Reading input file... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   Summary: 226363 sgRNAs, 18931 genes, 1 lists; skipped sgRNAs:0 \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   Computing lo-values for each group... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   Skipping gene NO-TARGET for permutation ... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   Warning: sgRNA  not found in the ranked list.  \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   Computing false discovery rate... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   Total # control sgRNAs: 1984 \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   Permuting genes with 2 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   Permuting genes with 4 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   Permuting genes with 6 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   Permuting genes with 8 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   Permuting genes with 9 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   Permuting genes with 10 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   Permuting genes with 11 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   Permuting genes with 12 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   Permuting genes with 24 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   Number of genes under FDR adjustment: 18931 \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   Saving to output file... \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   Suppressing the output of gene NO-TARGET since it is negative ontrol genes. \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:   RRA completed. \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17:    \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17: End command message. \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:17: Loading top 10 genes from data_out/CRISPRa.IFNG.CD8.orig.gene.low.txt: MAP4K1,SLA2,GRAP,MUC1,JMJD1C,LAT2,SETD1B,GATA3,EBF2,FOXF1 \n",
      "INFO  @ Wed, 09 Oct 2024 18:04:18: Loading top 10 genes from data_out/CRISPRa.IFNG.CD8.orig.gene.high.txt: VAV1,IFNG,CD28,TNFRSF1A,TNFRSF1B,TBX21,IL1R1,RELA,IL9R,FOXQ1 \n"
     ]
    }
   ],
   "source": [
    "! {mageck_command}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "CRISPR_type = 'CRISPRa'\n",
    "cell_types = 'CD4'\n",
    "cytokine = 'IFNG'\n",
    "study_type = 'supplemental'\n",
    "input_folder = 'input_data'\n",
    "count_filename = f'{input_folder}/{CRISPR_type}.{cell_types}_count.{study_type}.txt'\n",
    "nt_filename = f'{input_folder}/{CRISPR_type}.{cell_types}_NO-TARGET.{study_type}.txt'\n",
    "out_filename = f'data_out/{CRISPR_type}.{cytokine}.{cell_types}.{study_type}'\n",
    "donor_1 = 15\n",
    "donor_2 = 16\n",
    "mageck_command = get_mageck_command(\n",
    "    count_filename,\n",
    "    donor_1,\n",
    "    donor_2,\n",
    "    cytokine,\n",
    "    out_filename,\n",
    "    nt_filename\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  @ Wed, 09 Oct 2024 18:23:41: Parameters: /home/julian/miniconda3/envs/mageckenv2/bin/mageck test -k input_data/CRISPRa.CD4_count.supplemental.txt -t Donor15_IFNG_high,Donor16_IFNG_high -c Donor15_IFNG_low,Donor16_IFNG_low -n data_out/CRISPRa.IFNG.CD4.supplemental --norm-method none --paired --control-sgrna input_data/CRISPRa.CD4_NO-TARGET.supplemental.txt --keep-tmp \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:41: Welcome to MAGeCK v0.5.9.2. Command: test \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:41: Loading count table from input_data/CRISPRa.CD4_count.supplemental.txt  \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:41: Processing 1 lines.. \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:41: Processing 100001 lines.. \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:41: Loaded 113238 records. \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:41: Loading R template file: /home/julian/miniconda3/envs/mageckenv2/lib/python3.7/site-packages/mageck/plot_template.RTemplate. \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:41: Loading R template file: /home/julian/miniconda3/envs/mageckenv2/lib/python3.7/site-packages/mageck/plot_template_indvgene.RTemplate. \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:41: Loading Rnw template file: /home/julian/miniconda3/envs/mageckenv2/lib/python3.7/site-packages/mageck/plot_template.Rnw. \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:41: Treatment samples:Donor15_IFNG_high,Donor16_IFNG_high \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:41: Treatment sample index:8,10 \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:41: Control samples:Donor15_IFNG_low,Donor16_IFNG_low \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:41: Control sample index:9,11 \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:41: Final size factor: 1.0 1.0 1.0 1.0 \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:42: The updated negative control ID file is:data_out/CRISPRa.IFNG.CD4.supplemental.negctrl_ids.txt \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:43: Detecting outliers from variance estimation.. Avg read count:[168.75], Var: [24030.740563252013] \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:43: Skipping 1294 sgRNAs from variance calculation because of their extreme high counts (> 4* STD (mean counts) ). \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:43: Before RRA, 65 sgRNAs are removed with zero counts in both group(s). \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:43: Use qnorm to reversely calculate sgRNA scores ... \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:47: Running command: RRA -i data_out/CRISPRa.IFNG.CD4.supplemental.plow.txt -o data_out/CRISPRa.IFNG.CD4.supplemental.gene.low.txt -p 0.4786604891105114 --control data_out/CRISPRa.IFNG.CD4.supplemental.negctrl_ids.txt --skip-gene NA --skip-gene na  \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56: Command message: \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   Welcome to RRA v 0.5.9. \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   1985 control sequences loaded. \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   Skipping gene NA for permutation ... \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   Skipping gene na for permutation ... \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   Reading input file... \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   Summary: 226411 sgRNAs, 18931 genes, 1 lists; skipped sgRNAs:0 \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   Computing lo-values for each group... \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   Skipping gene NO-TARGET for permutation ... \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   Warning: sgRNA  not found in the ranked list.  \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   Computing false discovery rate... \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   Total # control sgRNAs: 1984 \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   Permuting genes with 2 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   Permuting genes with 4 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   Permuting genes with 6 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   Permuting genes with 8 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   Permuting genes with 10 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   Permuting genes with 11 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   Permuting genes with 12 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   Permuting genes with 24 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   Number of genes under FDR adjustment: 18931 \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   Saving to output file... \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   Suppressing the output of gene NO-TARGET since it is negative ontrol genes. \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:   RRA completed. \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56:    \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56: End command message. \n",
      "INFO  @ Wed, 09 Oct 2024 18:23:56: Running command: RRA -i data_out/CRISPRa.IFNG.CD4.supplemental.phigh.txt -o data_out/CRISPRa.IFNG.CD4.supplemental.gene.high.txt -p 0.21672533578315542 --control data_out/CRISPRa.IFNG.CD4.supplemental.negctrl_ids.txt --skip-gene NA --skip-gene na  \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03: Command message: \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   Welcome to RRA v 0.5.9. \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   1985 control sequences loaded. \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   Skipping gene NA for permutation ... \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   Skipping gene na for permutation ... \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   Reading input file... \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   Summary: 226411 sgRNAs, 18931 genes, 1 lists; skipped sgRNAs:0 \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   Computing lo-values for each group... \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   Skipping gene NO-TARGET for permutation ... \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   Warning: sgRNA  not found in the ranked list.  \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   Computing false discovery rate... \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   Total # control sgRNAs: 1984 \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   Permuting genes with 2 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   Permuting genes with 4 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   Permuting genes with 6 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   Permuting genes with 8 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   Permuting genes with 10 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   Permuting genes with 11 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   Permuting genes with 12 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   Permuting genes with 24 sgRNAs... \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   Number of genes under FDR adjustment: 18931 \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   Saving to output file... \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   Suppressing the output of gene NO-TARGET since it is negative ontrol genes. \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:   RRA completed. \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03:    \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03: End command message. \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:03: Loading top 10 genes from data_out/CRISPRa.IFNG.CD4.supplemental.gene.low.txt: MUC1,SPN,LAT2,MAP4K1,STAT6,FOXI3,SOX13,LYN,ZNF445,CD37 \n",
      "INFO  @ Wed, 09 Oct 2024 18:24:04: Loading top 10 genes from data_out/CRISPRa.IFNG.CD4.supplemental.gene.high.txt: IFNG,VAV1,MAML2,CD28,TNFRSF1A,TBX21,DEPDC7,BICDL2,TNFRSF1B,CD247 \n"
     ]
    }
   ],
   "source": [
    "! {mageck_command}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mageckenv2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
